Sources/BowEffects/Data/Internal/CancelablePromise.swift

```diff
@@ -0,0 +1,156 @@
+import Bow
+import Foundation
+
+internal class CancelablePromise<F: Concurrent, A: Equatable>: Promise<F, A> where F.E: Equatable & PromiseError {
+    private let state = Atomic<PromiseState<F.E, A>>(.pending(joiners: [:]))
+    
+    override func get() -> Kind<F, A> {
+        return F.defer {
+            switch self.state.value {
+            case let .complete(value: value): return F.pure(value)
+            case .pending: return F.cancelable { callback in
+                let id = self.unsafeRegister(callback)
+                return F.delay { self.unregister(id) }
+            }
+            case let .error(error): return F.raiseError(error)
+            }
+        }
+    }
+    
+    override func tryGet() -> Kind<F, Option<A>> {
+        return F.delay {
+            let current = self.state.value
+            switch current {
+            case let .complete(value: a): return .some(a)
+            default: return .none()
+            }
+        }
+    }
+    
+    override func complete(_ a: A) -> Kind<F, ()> {
+        return tryComplete(a).flatMap { didComplete in
+            if didComplete {
+                return F.pure(())
+            } else {
+                return F.raiseError(F.E.alreadyFulfilled)
+            }
+        }
+    }
+    
+    override func tryComplete(_ a: A) -> Kind<F, Bool> {
+        return F.defer { self.unsafeTryComplete(a) }
+    }
+    
+    override func error<E: Error>(_ e: E) -> Kind<F, ()> {
+        return tryError(e).flatMap { didError in
+            if didError {
+                return F.pure(())
+            } else {
+                return F.raiseError(F.E.alreadyFulfilled)
+            }
+        }
+    }
+    
+    override func tryError<E: Error>(_ e: E) -> Kind<F, Bool> {
+        if let error = e as? F.E {
+            return F.defer { self.unsafeTryError(error) }
+        } else {
+            return F.pure(false)
+        }
+    }
+    
+    private func unsafeTryComplete(_ a: A) -> Kind<F, Bool> {
+        let current = state.value
```

Comment:
**just a suggestion** I think if you use a `guard case let .pending(joiners: joiners) = state.value else { F.pure(false); return }` it is more readable :)

Reply:
Nice catch!